### PR TITLE
Implement DPatch evasion attack on object detectors

### LINKED ISSUE #276: Implement DPatch: An Adversarial Patch Attack on Object Detectors
Implement DPatch: An Adversarial Patch Attack on Object Detectors

Paper link: https://arxiv.org/abs/1806.02299v4

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

This pull request implements the evasion attack DPatch for object detectors.

Fixes #276 

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/estimators/object_detection/PyTorchFasterRCNN.py [hunk 1]
```diff
@@ -112,20 +124,26 @@ def loss_gradient(self, x, y, **kwargs):
...
+                y[i]["labels"] = torch.LongTensor(y_i["labels"]).to(self._device)
+                y[i]["scores"] = torch.Tensor(y_i["scores"]).to(self._device)
 
         transform = torchvision.transforms.Compose([torchvision.transforms.ToTensor()])
 
         image_tensor_list = list()
 
         for i in range(x.shape[0]):
-            img = transform(x[i])
+            img = transform(x[i] / self.clip_values[1]).to(self._device)
```
Comment (hesseltuinhof) [inline art/estimators/object_detection/PyTorchFasterRCNN.py:146]: I observe that on several occasions you divide by `clip_values[1]`. This would break when a user provides clip values equal to `(min_clip, 0)` where `min_clip<0`. Would something like `np.abs(clip_values).max()` or `np.abs(clip_values[1] - clip_values[0]) / 2` be more senisble?

File: art/estimators/object_detection/PyTorchFasterRCNN.py [hunk 2]
```diff
@@ -160,6 +180,8 @@ def loss_gradient(self, x, y, **kwargs):
 
         assert grads.shape == x.shape
 
+        grads = grads / self.clip_values[1]
```
Comment (hesseltuinhof) [inline art/estimators/object_detection/PyTorchFasterRCNN.py:183]: See the previous comment.

File: art/estimators/object_detection/PyTorchFasterRCNN.py [hunk 3]
```diff
@@ -184,14 +206,13 @@ def predict(self, x, **kwargs):
...
         x, _ = self._apply_preprocessing(x, y=None, fit=False)
-        x = x.astype(np.uint8)
 
         transform = torchvision.transforms.Compose([torchvision.transforms.ToTensor()])
 
         image_tensor_list = list()
 
         for i in range(x.shape[0]):
-            image_tensor_list.append(transform(x[i]))
+            image_tensor_list.append(transform(x[i] / self.clip_values[1]).to(self._device))
```
Comment (hesseltuinhof) [inline art/estimators/object_detection/PyTorchFasterRCNN.py:215]: See the previous comment.

File: art/attacks/evasion/dpatch.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+        :rtype: `np.ndarray`
+        """
+
+        assert (
+            x.shape[self.estimator.channel_index] == self.patch_shape[self.estimator.channel_index - 1]
+        ), "The color channel index of the images and the patch have to be identical."
+
+        assert y is None, "The DPatch attack does not use target labels."
+
+        assert len(x.shape) == 4, "The adversarial patch can only be applied to images."
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/dpatch.py]: That's great!

File: art/attacks/evasion/dpatch.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+        for i_step in range(self.max_iter):
+            print("i_step", i_step)
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/dpatch.py:106]: Yes, fixed in 37e2afc510c3f10aa8be5209d068fe5fc191b5d6

File: art/attacks/evasion/dpatch.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+
+            self._patch -= patch_gradients * self.learning_rate
+            self._patch = np.clip(self._patch, a_min=self.estimator.clip_values[0], a_max=self.estimator.clip_values[1])
+
+        return self._patch
+
+    @staticmethod
+    def _augment_images_with_patch(x, patch, random_location, channel_index):
+        """
+        Augment images with randomly shifted patch.
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/dpatch.py]: Correct, it's not always randomly. Fixed in 5ada7aa6023f1f8531eeaa51f8f6cc8e4eec0660

File: art/attacks/evasion/dpatch.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+
+        patched_images, _ = self._augment_images_with_patch(x=x, patch=patch_local, random_location=random_location)
+
+        return patched_images
+
+    def set_params(self, **kwargs):
+        """
+        Take in a dictionary of parameters and applies attack-specific checks before saving them as attributes.
+
+        :param patch_shape: The shape of the adversarial path as a tuple of shape (width, height, nb_channels).
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/dpatch.py]: Good catch, that was the actual intention, fixed in 981bdce1f69e19171df428b530d21d3bd1d1dfbb

File: art/attacks/evasion/dpatch.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+
+    def __init__(
+        self, estimator, patch_shape=(40, 40, 3), learning_rate=5.0, max_iter=500, batch_size=16,
+    ):
+        """
+        Create an instance of the :class:`.DPatch`.
+
+        :param estimator: A trained object detector.
+        :type estimator: :class:`.ObjectDetectorMixin`
+        :param patch_shape: The shape of the adversarial path as a tuple of shape (width, height, nb_channels).
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/dpatch.py]: Good catch, that was the actual intention, fixed in 149ecb1cea0f15fc4e2cedbe56525ed9da47d3a2

File: art/estimators/object_detection/PyTorchFasterRCNN.py [hunk 9]
```diff
@@ -139,6 +157,8 @@ def loss_gradient(self, x, y, **kwargs):
             else:
                 loss = loss + output[loss_name]
 
+        print("loss", loss)
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/estimators/object_detection/PyTorchFasterRCNN.py]: Fixed in 8913a5bded610c8b94fbd9b4da6d8024bb341b80

File: art/attacks/evasion/dpatch.py [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+            else:
+                i_x_1 = 0
+                i_y_1 = 0
+
+            i_x_2 = i_x_1 + patch.shape[0]
+            i_y_2 = i_y_1 + patch.shape[1]
+
+            transformations.append({"i_x_1": i_x_1, "i_y_1": i_y_1, "i_x_2": i_x_2, "i_y_2": i_y_2})
+
+            x_copy[i_image, i_x_1:i_x_2, i_y_1:i_y_2, :] = patch_copy
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/dpatch.py:199]: I think the existing code should work because we only update the documentation of the shape convention, it seems to me the indexing was always using the new shape convention.

File: art/attacks/evasion/dpatch.py [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+
+            if random_location:
+                i_x_1 = random.randint(0, x.shape[1] - 1 - patch.shape[0])
+                i_y_1 = random.randint(0, x.shape[2] - 1 - patch.shape[1])
+            else:
+                i_x_1 = 0
+                i_y_1 = 0
+
+            i_x_2 = i_x_1 + patch.shape[0]
+            i_y_2 = i_y_1 + patch.shape[1]
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/dpatch.py]: I agree, fixed in c7b7d6ee6f98aa42fd66096b8f6225dd1c39b6b2

File: art/attacks/evasion/dpatch.py [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+            x_copy = np.swapaxes(x_copy, 1, 3)
+            x_copy = np.swapaxes(x_copy, 2, 3)
+            patch_copy = np.swapaxes(patch_copy, 0, 2)
+            patch_copy = np.swapaxes(patch_copy, 1, 2)
+
+        for i_image in range(x.shape[0]):
+
+            if random_location:
+                i_x_1 = random.randint(0, x.shape[1] - 1 - patch.shape[0])
+                i_y_1 = random.randint(0, x.shape[2] - 1 - patch.shape[1])
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/dpatch.py]: I agree, fixed in c7b7d6ee6f98aa42fd66096b8f6225dd1c39b6b2

File: art/attacks/evasion/dpatch.py [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+        """
+        transformations = list()
+        x_copy = x.copy()
+        patch_copy = patch.copy()
+
+        if channel_index == 1:
+            x_copy = np.swapaxes(x_copy, 1, 3)
+            x_copy = np.swapaxes(x_copy, 2, 3)
+            patch_copy = np.swapaxes(patch_copy, 0, 2)
+            patch_copy = np.swapaxes(patch_copy, 1, 2)
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/dpatch.py]: I agree, fixed in c7b7d6ee6f98aa42fd66096b8f6225dd1c39b6b2

File: art/attacks/evasion/dpatch.py [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+            i_x_2 = i_x_1 + patch.shape[0]
+            i_y_2 = i_y_1 + patch.shape[1]
+
+            transformations.append({"i_x_1": i_x_1, "i_y_1": i_y_1, "i_x_2": i_x_2, "i_y_2": i_y_2})
+
+            x_copy[i_image, i_x_1:i_x_2, i_y_1:i_y_2, :] = patch_copy
+
+        if channel_index == 1:
+            x_copy = np.swapaxes(x_copy, 1, 3)
+            x_copy = np.swapaxes(x_copy, 2, 3)
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/dpatch.py]: I agree, fixed in c7b7d6ee6f98aa42fd66096b8f6225dd1c39b6b2

File: tests/attacks/evasion/test_dpatch.py [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+    n_train = 100
+    n_test = 11
+    yield (x_train_mnist[:n_train], y_train_mnist[:n_train], x_test_mnist[:n_test], y_test_mnist[:n_test])
+
+
+@pytest.mark.parametrize("random_location", [True, False])
+def test_augment_images_with_patch(random_location, fix_get_mnist_subset):
+    (x_train_mnist, y_train_mnist, x_test_mnist, y_test_mnist) = fix_get_mnist_subset
+
+    patch = np.ones(shape=(4, 4, 1)) * 0.5
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply tests/attacks/evasion/test_dpatch.py]: I agree, fixed in c7b7d6ee6f98aa42fd66096b8f6225dd1c39b6b2

File: art/estimators/object_detection/PyTorchFasterRCNN.py [hunk 16]
```diff
@@ -92,6 +92,18 @@ def __init__(
...
             self._model = model
+
+        # Set device
+        import torch
+
+        if torch.cuda.is_available():
+            cuda_idx = torch.cuda.current_device()
+            self._device = torch.device("cuda:{}".format(cuda_idx))
+        else:
+            self._device = torch.device("cpu")
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/estimators/object_detection/PyTorchFasterRCNN.py]: I agree and have changed it to be similar as in `PyTorchClassifier` with 4cb942792daf02d053bdc511e1183b5a1d5bf0dd

File: art/estimators/object_detection/PyTorchFasterRCNN.py (hunk 1 above)
Comment (beat-buesser) [reply art/estimators/object_detection/PyTorchFasterRCNN.py:152]: You are right. The reason for this is that this estimator is model-specific for PyTorch's FasterRCNN which expects images in the pixel range [0, 1]. To enable inputs to `PyTorchFasterRCNN` in the range [0, 255] it normalises with `clip_values[1]`. I have now added assertions to prevent unexpected `clip_values` where where `min_clip!=0`.

File: art/estimators/object_detection/PyTorchFasterRCNN.py (hunk 2 above)
Comment (beat-buesser) [reply art/estimators/object_detection/PyTorchFasterRCNN.py:187]: You are right. The reason for this is that this estimator is model-specific for PyTorch's FasterRCNN which expects images in the pixel range [0, 1]. To enable inputs to `PyTorchFasterRCNN` in the range [0, 255] it normalises with `clip_values[1]`. I have now added assertions to prevent unexpected `clip_values` where where `min_clip!=0`.

File: art/estimators/object_detection/PyTorchFasterRCNN.py (hunk 3 above)
Comment (beat-buesser) [reply art/estimators/object_detection/PyTorchFasterRCNN.py:219]: You are right. The reason for this is that this estimator is model-specific for PyTorch's FasterRCNN which expects images in the pixel range [0, 1]. To enable inputs to `PyTorchFasterRCNN` in the range [0, 255] it normalises with `clip_values[1]`. I have now added assertions to prevent unexpected `clip_values` where where `min_clip!=0`.

File: art/estimators/object_detection/PyTorchFasterRCNN.py (hunk 1 above)
Comment (hesseltuinhof) [reply art/estimators/object_detection/PyTorchFasterRCNN.py:152]: Awesome! Thanks for the clarification,  makes sense to me now +1:

File: art/attacks/evasion/dpatch.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (hesseltuinhof) [reply art/attacks/evasion/dpatch.py:197]: That makes sense :+1:  Agree!

**Comment by beat-buesser** [comment]
@hesseltuinhof Thank you very much for this great and thorough review!
